bandit/plugins/pytorch_load_save.py

```diff
@@ -0,0 +1,68 @@
+# Copyright (c) 2024 Stacklok, Inc.
+#
+# SPDX-License-Identifier: Apache-2.0
+r"""
+=========================================
+B704: Test for unsafe PyTorch load or save
+=========================================
+
+This plugin checks for the use of `torch.load` and `torch.save`. Using `torch.load`
+with untrusted data can lead to arbitrary code execution, and improper use of
+`torch.save` might expose sensitive data or lead to data corruption.
+
+:Example:
+
+.. code-block:: none
+
+        >> Issue: Use of unsafe PyTorch load or save
+        Severity: Medium   Confidence: High
+        CWE: CWE-94 (https://cwe.mitre.org/data/definitions/94.html)
+        Location: examples/pytorch_load_save.py:8
+        7    loaded_model.load_state_dict(torch.load('model_weights.pth'))
+        8    another_model.load_state_dict(torch.load('model_weights.pth', map_location='cpu'))
+        9
+        10   print("Model loaded successfully!")
+
+.. seealso::
+
+     - https://cwe.mitre.org/data/definitions/94.html
+
+.. versionadded:: 1.7.8
+
+"""
+import bandit
+from bandit.core import issue
+from bandit.core import test_properties as test
+
+
+@test.checks("Call")
+@test.test_id(
+    "B704"
+)  # Ensure the test ID is unique and does not conflict with existing Bandit tests
+def pytorch_load_save(context):
+    """
+    This plugin checks for the use of `torch.load` and `torch.save`. Using `torch.load`
+    with untrusted data can lead to arbitrary code execution, and improper use of
+    `torch.save` might expose sensitive data or lead to data corruption.
+    """
+    imported = context.is_module_imported_exact("torch")
+    qualname = context.call_function_name_qual
+    if not imported and isinstance(qualname, str):
+        return
+
+    qualname_list = qualname.split(".")
+    func = qualname_list[-1]
+    if all(
+        [
+            "torch" in qualname_list,
+            func in ["load"],
+            not context.check_call_arg_value("map_location", "cpu"),
+        ]
+    ):
+        return bandit.Issue(
+            severity=bandit.MEDIUM,
+            confidence=bandit.HIGH,
```

Comment:
Is there a safe version of these? If not, I wonder if we should lower the confidence as this may be the only way to load/save trusted data as well.

Reply:
Yes, I fear this might be a very noisy test if it is common to use load.

Reply:
I don't believe there is from research, most folks recommend perform a digest integrity check for load.

Reply:
Also, these inject type issues would be better detected via a taint analysis mechanism rather than just detect the call. As we get more of these, it can add more noise to the output if we don't improve the accuracy. The SQL inject plugin is doing a rudimentary backtrack form of taint analysis. We probably need to consider a larger solution to this sort of problem.